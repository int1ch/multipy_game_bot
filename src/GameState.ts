import {
  Game,
  PlainGameState as InternalGameState,
  Question,
} from "./GameStateShifter";

export type mSeconds = number;
//Обертка
export class GameState {
  constructor(protected gameState: InternalGameState) {}

  static newGame(game: Game) {
    const state: InternalGameState = {
      questions: game.questions,
      questionN: -1,
      failedQuestions: {},
      score: 0,
      try: 0,
      startTs: 0,
      startQuestionTs: 0,
    };
    return new GameState(state);
  }
  public startGame() {
    const state = this.gameState;
    state.questionN = 0;
    state.startTs = new Date().getTime();
    state.startQuestionTs = new Date().getTime();
    state.score = 0;
    state.try = 0;
  }

  public get score() {
    return this.gameState.score;
  }
  public gameElapsedTimeMs(): mSeconds {
    return new Date().getTime() - this.gameState.startTs;
  }
  public questionElapsedTimeMs(): mSeconds {
    if (!this.gameState.startQuestionTs) {
      console.log("Invalid state:", this.gameState);
      throw new Error("startQuestionTs not set!");
    }
    return new Date().getTime() - this.gameState.startQuestionTs;
  }
  public getCurrentQuestion() {
    const n = this.gameState.questionN;
    const question = this.gameState.questions[n];
    return question;
  }
  public isLastQuestion() {
    const n = this.gameState.questionN;
    const questions = this.gameState.questions;
    return n + 1 >= questions.length;
  }
  //state modifiers
  public getCurrentQuestionAttempts() {
    return this.gameState.try;
  }
  public increaseQuestionAttempts() {
    this.gameState.try++;
    //save session
  }
  public stateNextQuestion() {
    this.gameState.questionN++;
    this.gameState.startQuestionTs = new Date().getTime();
  }
  public addCurrentQuestionToFailed(question: Question) {
    this.gameState.failedQuestions[question.text] = question.answer;
  }
  public get failedQuestions() {
    return this.gameState.failedQuestions;
  }
  public increaseScore() {
    this.gameState.score++;
  }
  public ref() {
    return this.gameState;
  }
}
