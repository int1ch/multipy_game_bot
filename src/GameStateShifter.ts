import { GameTextGenerator } from "./GameTextGenerator";

export interface Question {
  text: string;
  answer: string;
  variants?: string[];
  wrongTimeout?: number;
  score?: number;
}
export interface Game {
  questions: Question[];
}

export interface PlainGameState {
  questionN: number;
  questions: Question[];
  failedQuestions: Record<string, string>;
  score: number;
  try: 0;
  startTs: number;
  startQuestionTs: number;
}

export interface QuestionStat {
  text: string;
  timeMs: number;
}
export interface GameStat {
  timeMs: number;
  score: number;
}
export interface ShifterResponse {
  text: string;
  variants?: string[];
  endGame?: boolean;
  result?: any; //FIXME
  stats?: {
    question?: QuestionStat;
    game?: GameStat;
  };
}
export interface Game {
  questions: Question[];
}

const GTG = GameTextGenerator;

export class GameStateShifter {
  protected MAX_TRYS = 2;
  protected state: PlainGameState;
  constructor(ctx: PlainGameState) {
    this.state = ctx;
  }
  public getState() {
    return this.state;
  }

  static newGame(game: Game) {
    const state: PlainGameState = {
      questions: game.questions,
      questionN: -1,
      failedQuestions: {},
      score: 0,
      try: 0,
      startTs: new Date().getTime(),
      startQuestionTs: 0,
    };
    return new GameStateShifter(state);
  }
  public checkAnswer(text: string): ShifterResponse {
    const gameState = this.state;
    const question = gameState.questions[gameState.questionN];
    if (text == question.answer) {
      gameState.score += question.score || 1;
      return this.nextQuestion();
    }
    return this.reQuestion();
  }
  public finishGame(): ShifterResponse {
    const endTime = new Date().getTime();
    const gameState = this.state;
    const startTime = gameState.startTs;
    let strTaken = "";

    let taken = 0;
    if (startTime) {
      taken = endTime - startTime;
      strTaken = (taken / 1000).toFixed(1);
      strTaken = `\nВаш Результат ${strTaken} секунд`;
    }
    let failedQuestionsArray: string[] = [];
    for (let [question, answer] of Object.entries(gameState.failedQuestions)) {
      failedQuestionsArray.push(question + " " + answer);
    }
    const failedQuestions = failedQuestionsArray.length
      ? "\n" + failedQuestionsArray.join("\n")
      : "";

    return {
      text: "Поздравляю игра закончена!" + strTaken + failedQuestions,
      endGame: true,
      result: {
        ms: taken,
        score: gameState.score,
      },
      stats: {
        game: {
          timeMs: taken,
          score: gameState.score,
        },
      },
    };
  }
  public startGame(): ShifterResponse {
    const gameState = this.state;
    gameState.questionN = -1;
    return this.nextQuestion();
  }
  public askQuestion(): ShifterResponse {
    const gameState = this.state;
    const nextQuestion = gameState.questions[gameState.questionN];
    //alter state
    return GTG.AskQuestion(nextQuestion);
  }
  static addQuestionStat(
    response: ShifterResponse,
    stats: undefined | QuestionStat
  ) {
    if (stats) {
      response.stats ||= {};
      response.stats!.question = stats;
    }
    return response;
  }
  public nextQuestion(): ShifterResponse {
    const gameState = this.state;
    const thisN = gameState.questionN;
    const nextN = gameState.questionN + 1;
    let stats: QuestionStat | undefined;
    if (thisN >= 0) {
      const elapsed = new Date().getTime() - gameState.startQuestionTs;
      stats = {
        text: gameState.questions[gameState.questionN].text,
        timeMs: elapsed,
      };
    }

    if (nextN >= gameState.questions.length) {
      return GameStateShifter.addQuestionStat(this.finishGame(), stats);
    }
    gameState.questionN = nextN;
    gameState.try = 0;
    gameState.startQuestionTs = new Date().getTime();

    return GameStateShifter.addQuestionStat(this.askQuestion(), stats);
  }
  public reQuestion(): ShifterResponse {
    const gameState = this.state;
    const question = gameState.questions[gameState.questionN];
    const answer = question.answer;

    gameState.failedQuestions[question.text] = answer;
    //try
    if (gameState.try > this.MAX_TRYS) {
      // FAIL
      const response = this.nextQuestion();
      response.text =
        `Снова не угадали, правильный ответ ${answer}\n` +
        "давайте слудующий вопрос\n" +
        response.text;

      return response;
    }
    // TRY AGAIN
    gameState.try++;
    return GTG.ReAskQuestion(question, gameState.try);
  }
}
