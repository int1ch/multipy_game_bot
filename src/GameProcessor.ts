import {
  Game,
  PlainGameState as InternalGameState,
  Question,
} from "./GameStateShifter";
import { GameTextGenerator, TextResponse } from "./GameTextGenerator";
import { GameContext } from "./gameRouter";
import { GameStats } from "./db/GameStats";
import central from "./CentralStation";
import { ReplyKeyboardMarkup, ReplyKeyboardRemove } from "@grammyjs/types";
import { Keyboard } from "grammy";
import { GameState } from "./GameState";

export class GameProcessor {
  protected MAX_ATTEMPTS = 3;
  protected _gameStats: GameStats | undefined;

  constructor(
    protected gameState: GameState,
    protected playerId: number,
    injection?: { gameStats: GameStats }
  ) {
    this._gameStats = injection?.gameStats;
  }
  static newGame(game: Game, playerId: number) {
    const newState = GameState.newGame(game);
    const processor = new GameProcessor(newState, playerId);
    return processor;
  }
  //DIjection for tests?
  get gameStats(): GameStats {
    if (this._gameStats) {
      return this._gameStats;
    }
    return central.gameStats;
  }
  public stateRef() {
    return this.gameState.ref();
  }
  public async startGame(): Promise<GameResponse> {
    this.gameState.startGame();
    return await this.askQuestion();
  }
  public async checkAnswer(text: string): Promise<GameResponse> {
    const question = this.gameState.getCurrentQuestion();
    //todo prepare text
    if (question.answer === text) {
      //save question stat here
      //save game stat sovewere in here
      return await this.moveToNextQuestionAndReturnStats({ scored: true });
    }
    this.gameState.addCurrentQuestionToFailed(question);
    return await this.reAskQuestion();
  }
  public async moveToNextQuestionAndReturnStats(options?: {
    previousQuestionFailed?: boolean;
    scored?: boolean;
  }): Promise<GameResponse> {
    const previousQuestionFailed = !!options?.previousQuestionFailed;

    const oldQuestion = this.gameState.getCurrentQuestion();
    const elapsedMs = this.gameState.questionElapsedTimeMs();
    const questionStat: QuestionStat = {
      text: oldQuestion.text,
      timeMs: elapsedMs,
      failed: previousQuestionFailed,
    };

    const response = await this.moveToNextQuestion(options);

    const result = {
      ...response,
      questionStat,
    };
    console.log("RESPONSE: ", response, result);
    return result;
  }
  public async moveToNextQuestion(options?: {
    previousQuestionFailed?: boolean;
    scored?: boolean;
  }): Promise<GameResponse> {
    const previousQuestionFailed = !!options?.previousQuestionFailed;
    const questionScored = !!options?.scored;

    if (questionScored) {
      this.gameState.increaseScore();
    }

    if (this.gameState.isLastQuestion()) {
      return await this.finishGame({ previousQuestionFailed });
    }

    const oldQuestion = this.gameState.getCurrentQuestion();
    this.gameState.stateNextQuestion();

    const nextQuestion = this.gameState.getCurrentQuestion();
    const textResponse = previousQuestionFailed
      ? GameTextGenerator.FailedThenNextQuestion(oldQuestion, nextQuestion)
      : GameTextGenerator.AskQuestion(nextQuestion);

    return textResponse;
  }
  public async finishGame(options?: {
    previousQuestionFailed: boolean;
  }): Promise<GameResponse> {
    const previousQuestionFailed = options?.previousQuestionFailed;
    const timeTakenMs = this.gameState.gameElapsedTimeMs();
    const score = this.gameState.score;

    //const telegamId = this.getTelegramId();
    const statsFetcher = null;
    const textResponse = await GameTextGenerator.FinishGame(
      timeTakenMs,
      score,
      this.gameState.failedQuestions,
      this.playerId,
      this.gameStats
    );

    return {
      ...textResponse,
      removeKeyboard: true,
      gameStat: {
        timeMs: timeTakenMs,
        score: score,
      },
      endGame: true,
    };
  }
  public async reAskQuestion() {
    const attempts = this.gameState.getCurrentQuestionAttempts();
    if (attempts >= this.MAX_ATTEMPTS) {
      return await this.moveToNextQuestionAndReturnStats({
        previousQuestionFailed: true,
      });
    }
    this.gameState.increaseQuestionAttempts();
    return this.askQuestion({ previousQuestionFailed: true });
  }
  public askQuestion(options?: { previousQuestionFailed: boolean }) {
    const previousQuestionFailed = options?.previousQuestionFailed;
    const question = this.gameState.getCurrentQuestion();
    const response = previousQuestionFailed
      ? GameTextGenerator.ReAskQuestion(
          question,
          this.gameState.getCurrentQuestionAttempts()
        )
      : GameTextGenerator.AskQuestion(question);

    return response;
  }
}
interface GameStat {
  timeMs: number;
  score: number;
}
interface QuestionStat {
  text: string;
  timeMs: number;
  failed?: boolean;
}
export interface GameResponse extends TextResponse {
  //text
  //variants
  removeKeyboard?: boolean;
  questionStat?: QuestionStat;
  gameStat?: GameStat;
  endGame?: boolean;
}
