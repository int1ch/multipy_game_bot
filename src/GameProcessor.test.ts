import exp from "constants";
import central from "./CentralStation";
import { GameProcessor } from "./GameProcessor";
import { GameState } from "./GameState";
import { PlainGameState } from "./GameStateShifter";
afterAll(() => {
  central.close();
});

const threeQuestionGame = {
  questions: [
    { text: "a", answer: "a", variants: ["a", "b"] },
    { text: "b", answer: "b", variants: ["a", "b"] },
    { text: "c", answer: "c" },
  ],
};

const playerId = 1;
const fixedTs = 1234567890;
function jClone<T>(x: T): T {
  return JSON.parse(JSON.stringify(x));
}

expect.extend({
  toBeTimestamp(received) {
    const t = typeof received;
    if (t !== "number") {
      return {
        message: () =>
          `expected ${received} to be a number(got ${t}) with milliseconds timestamp `,
        pass: false,
      };
    }
    const ts = new Date().getTime();
    const pass = received > ts - 10000 && received <= ts;
    if (pass) {
      return {
        message: () =>
          `not expected ${received} to be close with current ts:${ts}`,
        pass: true,
      };
    }
    return {
      message: () => `expected ${received} to be close with current ts:${ts}`,
      pass: false,
    };
  },
});
declare global {
  namespace jest {
    interface Matchers<R> {
      toBeTimestamp(): R;
    }
  }
}

describe("3 question game", () => {
  let state!: PlainGameState;
  it("start", async () => {
    const processor = GameProcessor.newGame(threeQuestionGame, playerId);
    const response = await processor.startGame();
    state = processor.stateRef();

    expect(response.text).toBe("a");

    expect(state).toBeTruthy();
    expect(state.questionN).toBe(0);
    expect(state.questions).toHaveLength(3);
    expect(state.startQuestionTs).toBeTimestamp();
    expect(state.startTs).toBeTimestamp();
    const clone = jClone(state);
    clone.startQuestionTs = fixedTs;
    clone.startTs = fixedTs;
    expect(clone).toMatchSnapshot();
  });
  it("ask q2", async () => {
    const gameState = new GameState(state);
    const processor = new GameProcessor(gameState, playerId);
    const response = await processor.checkAnswer("a");

    const clone = jClone(state);
    clone.startQuestionTs = fixedTs;
    clone.startTs = fixedTs;
    expect(clone).toMatchSnapshot();

    expect(response.questionStat?.timeMs).not.toBeTimestamp();
    response.questionStat!.timeMs = 1;

    expect(response.text).toBe("b");
    expect(response.variants).toHaveLength(2);
    expect(response.questionStat).toBeTruthy();
    expect(response).toMatchSnapshot();
  });
  it("ask q3", async () => {
    const gameState = new GameState(state);
    const processor = new GameProcessor(gameState, playerId);
    const response = await processor.checkAnswer("b");
    state = processor.stateRef();

    expect(response.text).toBe("c");
    const clone = jClone(state);
    clone.startQuestionTs = fixedTs;
    clone.startTs = fixedTs;
    expect(clone).toMatchSnapshot();
  });
  it("finish", async () => {
    const gameState = new GameState(state);
    expect(gameState.isLastQuestion()).toBeTruthy();

    const processor = new GameProcessor(gameState, playerId);
    const response = await processor.checkAnswer("c");
    state = processor.stateRef();

    expect(response.gameStat).toBeTruthy();
    expect(response.gameStat?.score).toBe(3);
    expect(response.gameStat?.timeMs).not.toBeTimestamp();
    expect(response.endGame).toBeTruthy();
    expect(response.removeKeyboard).toBeTruthy();

    expect(state).toBeTruthy();
    expect(state.startQuestionTs).toBeTimestamp();
    expect(state.startTs).toBeTimestamp();
    const clone = jClone(state);
    clone.startQuestionTs = fixedTs;
    clone.startTs = fixedTs;
    expect(clone).toMatchSnapshot();
  });
});

describe("move to nextQuestion", () => {
  async function newGame() {
    const processor = GameProcessor.newGame(threeQuestionGame, playerId);
    const response = await processor.startGame();
    return processor;
  }

  test("normal way", async () => {
    const processor = await newGame();
    const response = await processor.moveToNextQuestion();
    expect(response.text).toBe("b");
  });
  test("failed way", async () => {
    const processor = await newGame();
    const response = await processor.moveToNextQuestion({
      previousQuestionFailed: true,
    });
    expect(response.text).toContain("b");
  });
});
