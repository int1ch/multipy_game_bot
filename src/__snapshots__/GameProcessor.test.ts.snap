// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`3 question game ask q2 1`] = `
Object {
  "failedQuestions": Object {},
  "questionN": 1,
  "questions": Array [
    Object {
      "answer": "a",
      "text": "a",
      "variants": Array [
        "a",
        "b",
      ],
    },
    Object {
      "answer": "b",
      "text": "b",
      "variants": Array [
        "a",
        "b",
      ],
    },
    Object {
      "answer": "c",
      "text": "c",
    },
  ],
  "score": 1,
  "startQuestionTs": 1234567890,
  "startTs": 1234567890,
  "try": 0,
}
`;

exports[`3 question game ask q2 2`] = `
Object {
  "questionStat": Object {
    "failed": false,
    "text": "a",
    "timeMs": 1,
  },
  "text": "b",
  "variants": Array [
    "a",
    "b",
  ],
}
`;

exports[`3 question game ask q3 1`] = `
Object {
  "failedQuestions": Object {},
  "questionN": 2,
  "questions": Array [
    Object {
      "answer": "a",
      "text": "a",
      "variants": Array [
        "a",
        "b",
      ],
    },
    Object {
      "answer": "b",
      "text": "b",
      "variants": Array [
        "a",
        "b",
      ],
    },
    Object {
      "answer": "c",
      "text": "c",
    },
  ],
  "score": 2,
  "startQuestionTs": 1234567890,
  "startTs": 1234567890,
  "try": 0,
}
`;

exports[`3 question game finish 1`] = `
Object {
  "failedQuestions": Object {},
  "questionN": 2,
  "questions": Array [
    Object {
      "answer": "a",
      "text": "a",
      "variants": Array [
        "a",
        "b",
      ],
    },
    Object {
      "answer": "b",
      "text": "b",
      "variants": Array [
        "a",
        "b",
      ],
    },
    Object {
      "answer": "c",
      "text": "c",
    },
  ],
  "score": 3,
  "startQuestionTs": 1234567890,
  "startTs": 1234567890,
  "try": 0,
}
`;

exports[`3 question game start 1`] = `
Object {
  "failedQuestions": Object {},
  "questionN": 0,
  "questions": Array [
    Object {
      "answer": "a",
      "text": "a",
      "variants": Array [
        "a",
        "b",
      ],
    },
    Object {
      "answer": "b",
      "text": "b",
      "variants": Array [
        "a",
        "b",
      ],
    },
    Object {
      "answer": "c",
      "text": "c",
    },
  ],
  "score": 0,
  "startQuestionTs": 1234567890,
  "startTs": 1234567890,
  "try": 0,
}
`;
